vars/buildDockerAndPublishImage.groovy

```diff
@@ -1,176 +1,86 @@
-import java.text.SimpleDateFormat;  
-import java.util.Date;
 
-def call(String imageName, Map config=[:]) {
-  if (!config.registry) {
-    if (infra.isTrusted() || infra.isInfra()) {
-      config.registry = "jenkinsciinfra/"
-    } else {
-      config.registry = "jenkins4eval/"
-    }
-  }
+import io.jenkins.infra.DockerConfig
+import io.jenkins.infra.InfraConfig
 
-  if (!config.dockerfile) {
-    config.dockerfile = "Dockerfile"
-  }
+def call(String imageName, Map config=[:]) {
 
-  if (!config.credentials) {
-    config.credentials = "jenkins-dockerhub"
-  }
+  // Initialize the groovy context
+  final DockerConfig dockerConfig = new DockerConfig(imageName, config, new InfraConfig(env))
 
-  if (!config.mainBranch) {
-    config.mainBranch = "master"
-  }
+  // Retrieve Library's Static File Resources
+  final String makefileContent = libraryResource 'io/jenkins/infra/docker/Makefile'
+  final String podTemplate = libraryResource 'io/jenkins/infra/docker/pod-template.yml'
 
   pipeline {
```

Comment:
I do not recommend trying to use Declarative syntax inside a library.

Reply:
The reasons behing keeping the declarative syntax here are because the effort to switch to scripted are crazy, and contribution gate should be reachable for a non pipeline expert.

Reading your comment, what would be the rationale behing using scripted pipeline (so we can assess the requirements here and see if it a reachable goal)?

Thanks for the review and the time taken on this subject !

Reply:
> what would be the rationale behing using scripted pipeline

Just trying to keep Declarative to its original goal—scriptless, fully structured project definitions suitable for round-tripping.

Reply:
Interesting, I wasn't really aware of this original goal. Thanks for sharing!
To be honest, I feel like there is an "intermediate" situation (very well represented by this particular function) 
where the initial repository was using a declarative, and the need to "reuse" it (whole) arised.

The goal of this PR (despite the branch name and some back and forth with scripted) is to add a stage for "testing",
which does not feel like to migrate the whole pipeline in scripted. (I admit that I tried initially, but in the end there was no real need to only call the "build" or the "test" step alone: either you call the whole pipeline, or not.

=> I've seen mostly this kind of need in GitLab' Library or in YAML-powered CI (using YAML anchors) which bias me in the "I don't want to deal with a full fledge scripted language as I already am dealing with shell,Make, gradle, etc.".

Reply:
I think @jglick  is a very good recommendation and if you are not planning to switch to a full scripted pipeline, I guess it worthwhile to document it about the why? People get inspiration from this public shared-library.

Reply:
Fair point, I'm only wondering where should be document such thing? I was thinking about adding comment on the top of the groovy file defining the function. Is there another location for this?